File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章砍竹子.py
#砍竹子 2022年第十三届省赛C/C++大学B组
#lanqiaoOJ 2117

from math import * 

f = [[0]*10 for _ in range(200010)] #二维数组初始化
stk = [0]*10                        #一维数组初始化
n = int(input())
a = list(map(int,input().split()))
ans = 0
for i in range(n):
    x = a[i]; top = 0
    while x>1:
        top += 1; stk[top] = x
        x = floor(sqrt(floor(x/2)+1))
    ans += top
    k = top; j = 0
    while k>0:  f[i][j] = stk[k]; k -= 1;j += 1
for j in range(10):
    for i in range(1,n):
        if f[i][j]>0 and f[i][j] == f[i-1][j]: ans -= 1
print(ans)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章砍竹子.java
//砍竹子 2022年第十三届省赛C/C++大学B组
//lanqiaoOJ 2117

import java.util.*;
public class Main{
    static Scanner cin = new Scanner(System.in);
    static long [][] f = new long[200010][10];
    static long [] stk = new long[10];
    public static void main(String[] args){
        int n = cin.nextInt() , ans = 0;
        for(int i = 0 ; i < n ; i ++){
            long x = cin.nextLong() , top = 0;
            while(x > 1){
                stk[(int) ++ top] = x;
                x = (long) Math.sqrt(x / 2 + 1);
            }
            ans += top;
            int k = (int) top;
            for(int j = 0 ; k >= 0 ; j ++){
                f[i][j] = stk[k];
                k --;
            }
        }
        for(int j = 0 ; j < 10 ; j ++) {
            for(int i = 1 ; i < n ; i ++){
                if(f[i][j] > 0 && f[i][j] == f[i - 1][j]) ans --;
            }
        }
        System.out.println(ans);
    }
}